# 1. [RI/2025/PRELIM/P2/Q3]
The school has been experiencing an increasing number of lost items. You are tasked to create a **Lost and Found** web application and store the data about lost items in a MongoDB database. Your app will allow the Student Affairs staff to submit a new lost item report, view existing lost items in the database, and perform a search on lost items database.


## Task 1 
Write a program to: 
- create a Mongo client to connect to database `lostfoundDB` 
- use the data in `lost_items.json` to populate the collection `reports` 
 
Save your program code as

```text
Task3_1_<name>_<index number>.py 
```

in a folder named `Task3_1_<name>_<index number>` 
<div style="text-align: right">[3]</div>


In [18]:
# Your Answer Here
from pymongo import MongoClient
import json

client = MongoClient('localhost', 27017)
db = client['lostfoundDB']
coll = db['reports']

coll.delete_many({})

with open('./resources/lost_items.json') as f:
    data = json.load(f)
    coll.insert_many(data)

## Task 2 
Write a Python program and the necessary files to create a web application. When either the route `"/reports"` or `"/"` is visited, the web application will return a HTML page to display a report table of all the lost items found in the database, under appropriate table headings, sorting the records by date in ascending order. 
 
Save your program code as 

```text
Task3_2_<name>_<index number>.py 
```

with any additional files / subfolders as needed in a folder named 

```text
Task3_2_<name>_<index number> 
```

Run the web application and save the returned reports page as 

```text
Task3_2_<name>_<index number>.html 
```
<div style="text-align: right">[5]</div>


In [1]:
# Your Answer Here
from flask import Flask, render_template
from pymongo import MongoClient

app = Flask(__name__, template_folder='./Task3_2')

@app.route('/')
def index():
    client = MongoClient('localhost', 27017)
    db = client['lostfoundDB']
    coll = db['reports']

    data=sorted(coll.find(), key=lambda x: x['date_lost'], reverse=True)

    return render_template('index.html', data=data)

app.run()

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit


## Task 3 
Implement the following functionality to allow for new lost items to be reported. 

Student Affairs staff will visit the `"/submit"` route to save information of new lost item to the database.

Create a `"Submit Lost Item"` HTML form that allows the staff to key in details like `item_name`, `description`, `category`, `date_lost` and `location`. 

You should do validation of the `date_lost`, and ensure that invalid dates (for example, `'2025-02-31'` and `'2025/12/12'`) are not accepted, and notify the user that the entered date is invalid. 

After submitting the completed form, a new document should be created in the `reports` collection, and the user should be redirected to the `/reports` URL, which will show the new lost item reported, along with existing lost items.

Save your program code as 

```text
Task3_3_<name>_<index number>.py 
```

with any additional files / subfolders as needed in a folder named

```text
Task3_3_<name>_<index number> 
```

Run the web application, add a new report as below, and save the resultant report page as 

```text
Task3_3_<name>_<index number>.html 
```

- Item Name: `Spectacles` 
- Description: `Pink spectacles left outside the Hall` 
- Category: `Accessories` 
- Date Lost: `2025-07-31` 
- Location: `Hall` 
<div style="text-align: right">[7]</div>


In [20]:
# Your Answer Here
from flask import Flask, render_template, flash, request, redirect, url_for
from pymongo import MongoClient
from datetime import datetime

app = Flask(__name__, template_folder='./Task3_2')
app.secret_key = 'meow'

@app.route('/')
def index():
    client = MongoClient('localhost', 27017)
    db = client['lostfoundDB']
    coll = db['reports']

    data=sorted(coll.find(), key=lambda x: x['date_lost'], reverse=True)

    return render_template('index.html', data=data)

@app.route('/submit', methods=["GET", "POST"])
def submit():
    if request.method == "GET":
        return render_template('submit.html')
    
    name = request.form.get('name')
    desc = request.form.get('description')
    cat = request.form.get('category')
    date = request.form.get('date')
    location = request.form.get('location')

    try:
        test_date = datetime.strptime(date, '%Y-%m-%d')
    except:
        flash('Invalid Date Lost')
        return render_template('submit.html')
    
    client = MongoClient('localhost', 27017)
    db = client['lostfoundDB']
    coll = db['reports']

    data = {'item_name': name, 'description': desc, 'category': cat, 'date_lost': date, 'location': location, 'recovered': False}

    coll.insert_one(data)
    return redirect(url_for('index'))

app.run()

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [22/Jun/2026 15:51:06] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [22/Jun/2026 15:51:10] "GET /submit HTTP/1.1" 200 -
127.0.0.1 - - [22/Jun/2026 15:51:18] "POST /submit HTTP/1.1" 302 -
127.0.0.1 - - [22/Jun/2026 15:51:18] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [22/Jun/2026 15:51:21] "GET / HTTP/1.1" 200 -


## Task 4 
Implement the following functionality to allow the staff to search for lost items 
- by category, or 
- by date, or 
- by both category and date. 

Student Affairs staff will visit the `"/search"` route to perform a search on lost items in the database.

Create a `"Search Lost Item"` HTML form that contains two form inputs, `"Category"` and `"Date Lost"`. If the staff to key in only one of the two fields, the search will only show items that fulfil the search field entered. If both fields are entered, only items that fulfil both criteria will be shown in a html table below the search form. 

Save your program code as 

```text
Task3_4_<name>_<index number>.py 
```

with any additional files / subfolders as needed in a folder named

```text
Task3_4_<name>_<index number> 
```

Search for lost items on date: `'2025-07-31'` and save output as 

```text
Task3_4_1_<name>_<index number>.html 
```

Search for lost items with 
- date: `'2025-07-31'` and 
- category: `"Accessories"` 
and save the output as 

```text
Task3_4_2_<name>_<index number>.html 
```
<div style="text-align: right">[7]</div>


In [23]:
# Your Answer Here
from flask import Flask, render_template, flash, request, redirect, url_for
from pymongo import MongoClient
from datetime import datetime

app = Flask(__name__, template_folder='./Task3_2')
app.secret_key = 'meow'

@app.route('/')
def index():
    client = MongoClient('localhost', 27017)
    db = client['lostfoundDB']
    coll = db['reports']

    data=sorted(coll.find(), key=lambda x: x['date_lost'], reverse=True)

    return render_template('index.html', data=data)

@app.route('/search', methods=["GET", "POST"])
def search():
    if request.method == "GET":
        return render_template('search.html')

    client = MongoClient('localhost', 27017)
    db = client['lostfoundDB']
    coll = db['reports']

    cat = request.form.get('category')
    date = request.form.get('date')

    query = {}
    if cat:
        query['category'] = cat
    if date:
        query['date_lost'] = date

    result = coll.find(query)

    return render_template('search.html', result=result)

@app.route('/submit', methods=["GET", "POST"])
def submit():
    if request.method == "GET":
        return render_template('submit.html')
    
    name = request.form.get('name')
    desc = request.form.get('description')
    cat = request.form.get('category')
    date = request.form.get('date')
    location = request.form.get('location')

    try:
        test_date = datetime.strptime(date, '%Y-%m-%d')
    except:
        flash('Invalid Date Lost')
        return render_template('submit.html')
    
    client = MongoClient('localhost', 27017)
    db = client['lostfoundDB']
    coll = db['reports']

    data = {'item_name': name, 'description': desc, 'category': cat, 'date_lost': date, 'location': location, 'recovered': False}

    coll.insert_one(data)
    return redirect(url_for('index'))

app.run()

 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [22/Jun/2026 16:03:27] "GET /search HTTP/1.1" 200 -
127.0.0.1 - - [22/Jun/2026 16:03:28] "POST /search HTTP/1.1" 200 -
127.0.0.1 - - [22/Jun/2026 16:03:30] "POST /search HTTP/1.1" 200 -
127.0.0.1 - - [22/Jun/2026 16:03:34] "POST /search HTTP/1.1" 200 -



---

# 2. [ASRJC/2025/PRELIM/P2/Q3]

A H2 Computing teacher has a collection of fill-in-the-blanks problems that needs to be managed using a MongoDB database.


## Task 1

The file `problems.txt` contains problems separated by an empty line.

Write program code to extract the data from the file and store them in a MongoDB collection named `problems` within a database named `computing`.

Note that the answers to the blanks in a problem should be stored as a list of strings, and the attempt count stored as an integer value.
<div style="text-align: right">[5]</div>


In [2]:
# Your Answer Here
from pymongo import MongoClient

client = MongoClient('localhost', 27017)
db = client['computing']
coll = db['problems']

with open('./resources/problems.txt') as f:
    data = [{}]
    for line in [i.strip() for i in f.readlines()]:
        if line == '':
            data.append({})
            continue
            
        value = line.split(': ')
        data[-1][value[0]] = int(value[1]) if value[1].isdigit() else value[1]

coll.insert_many(data)

print(coll.find())

## Task 2

Write MongoDB query logic to extract problems that meet both of the following criteria:

 * the topic is either `‘Data Structures’` or `‘Algorithms’`
 * the number of attempts for the problems is between 5 and 10 (both inclusive).

Display the question texts of these matching problems. <div style="text-align: right">[3]</div>


In [7]:
# Your Answer Here
from pymongo import MongoClient

client = MongoClient('localhost', 27017)
db = client['computing']
coll = db['problems']

problems = coll.find({'Topic': {'$in': ['Data Structures', 'Algorithms']}, 
                      'Attempts': {'$gte': 5, '$lte': 10}
                      })

for p in problems:
    print(p['Question'])

A memory l___ occurs when dynamically allocated memory is not properly f___. This causes programs to consume increasing amounts of m___.
In a binary search tree, all values in the left subtree are s___ than the root, and all values in the right subtree are l___.
Bubble sort compares a___ elements and s___ them if they are in wrong order.
Merge sort uses d___ and c___ strategy, recursively splitting arrays then m___ sorted subarrays back together.
A stack follows the L___ I___ F___ O___ principle where the last element added is the first one r___.


## Task 3

Since the topic `‘Socket Programming’` will no longer be there in the revised H2 Computing syllabus, the teacher wants to update the collection by removing all problems tagged with that topic.

Write MongoDB logic to perform this update and execute it.
<div style="text-align: right">[2]</div>


In [8]:
# Your Answer Here
from pymongo import MongoClient

client = MongoClient('localhost', 27017)
db = client['computing']
coll = db['problems']

coll.delete_many({'Topic': 'Socket Programming'})

DeleteResult({'n': 3, 'ok': 1.0}, acknowledged=True)

## Task 4

Write program code that:

 * randomly selects a problem from the MongoDB collection
 * displays the topic and the question text of the selected problem
 * prompts the student to enter an answer for each blank in the question, reminding him to include the first letter in his response
 * checks the student’s answers against the expected answers (case-insensitive)
 * outputs summative results (e.g. “You got 3 out of 5 answers correct\!”), without revealing what the correct answers are
 * increments the attempt count for that problem by 1 in the database.

If the student’s answers to the problem are not all correct, prompt the student to enter all the answers again for the same problem.

This continues until the student answers the entire problem correctly.

Test your code appropriately.
<div style="text-align: right">[7]</div>


In [ ]:
# Your Answer Here
from pymongo import MongoClient
import random

client = MongoClient('localhost', 27017)
db = client['computing']
coll = db['problems']

problems = list(coll.find())
selected = random.choice(problems)

prev = None
blanks = 0
for char in selected['Question']:
    if char == '_' and prev != '_':
        blanks +=1
    prev = char

print(f"Topic: {selected['Topic']}\nQuestion: {selected['Question']}")

n_correct = 0
while n_correct < blanks:
    n_correct = 0
    for i in range(blanks):
        ans = input(f'What is your ans for blank {i+1}')
        print(ans, selected['Answers'].split(',')[i])
        if ans.lower() == selected['Answers'].split(',')[i].lower():
            n_correct += 1

    print(f'You got {n_correct}/{blanks} correct!')
    coll.update_one({'_id': selected['_id']}, {'$inc': {'Attempts': 1}})

Topic: Algorithms
Question: Binary search requires the array to be s___ beforehand and works by repeatedly h____ the search space.
 sorted
sad halving
You got 0/2 correct!
d sorted
sorted halving
You got 0/2 correct!
sorted sorted
answer halving
You got 1/2 correct!



---

# 3. [RI/2024/PRELIM/P2/Q1]
An online shopping platform stores its data in a MongoDB database.

The platform needs to store information about products, customers, and orders. 

Your task is to perform several operations on the database.


## Task 1
Write Python code to create a MongoDB database named `online_shop`. Ensure that this database is empty before proceeding.

Three JSON files are provided, each containing multiple records to be inserted into specific collections within the database. The files are `PRODUCTS.json`, `CUSTOMERS.json` and `ORDERS.json`.

Create collections named `products`, `customers`, and `orders`, and insert the data from each JSON file into the corresponding collection: <div style="text-align: right">[3]</div> 

<center>
<img src="img/2024_ri_p2_q1_img_1.png" width="600" align="center"/>
</center>

<center>
<img src="img/2024_ri_p2_q1_img_2.png" width="600" align="center"/>
</center>


In [2]:
# Your Answer Here
from pymongo import MongoClient
import json

client = MongoClient('localhost', 27017)
db = client['online_shop']

prod_coll = db['products']
cust_coll = db['customers']
orders_coll = db['orders']

with open('./resources/PRODUCTS.json') as f:
    prod_coll.insert_many(json.load(f))

with open('./resources/CUSTOMERS.json') as f:
    cust_coll.insert_many(json.load(f))

with open('./resources/ORDERS.json') as f:
    orders_coll.insert_many(json.load(f))

## Task 2
Add to your program code to retrieve and display all products in the `Books` category from the `products` collection.

Use the `pprint` module to display your data in a readable format:
>```
>from pprint import pprint 
>pprint(your_data)
>```
Run your program. <div style="text-align: right">[2]</div>


In [4]:
# Your Answer Here
from pprint import pprint

client = MongoClient('localhost', 27017)
db = client['online_shop']

prod_coll = db['products']

books = list(prod_coll.find({'category': 'Books'}))
pprint(books)

[{'_id': ObjectId('6a3b4c05ac03d6eda1d9ab5a'),
  'category': 'Books',
  'model': 'Classic Edition',
  'name': 'Novel: The Great Gatsby',
  'price': 15,
  'product_id': 11,
  'stock': 50},
 {'_id': ObjectId('6a3b4c05ac03d6eda1d9ab5b'),
  'category': 'Books',
  'model': '2020 Edition',
  'name': 'Cookbook: The Joy of Cooking',
  'price': 30,
  'product_id': 12,
  'stock': 30}]


## Task 3
Add to your program code to find all orders placed by the customer named `Charles`.
Assume there is only one customer with this name.

Display the details of these orders from the orders collection.
Use `pprint` to display your data in a readable format.
Run your program. <div style="text-align: right">[4]</div>


In [7]:
# Your Answer Here
client = MongoClient('localhost', 27017)
db = client['online_shop']

cust_coll = db['customers']
orders_coll = db['orders']

cust_id = cust_coll.find_one({'name': 'Charles'})['customer_id']
orders = list(orders_coll.find({'customer_id': cust_id}))

pprint(orders)

[{'_id': ObjectId('6a3b4c05ac03d6eda1d9ab64'),
  'customer_id': 3,
  'order_date': '2024-04-22',
  'order_id': 4,
  'products': [{'product_id': 5, 'quantity': 1},
               {'product_id': 2, 'quantity': 1},
               {'product_id': 6, 'quantity': 1}]},
 {'_id': ObjectId('6a3b4c05ac03d6eda1d9ab66'),
  'customer_id': 3,
  'order_date': '2024-06-18',
  'order_id': 6,
  'products': [{'product_id': 2, 'quantity': 1},
               {'product_id': 1, 'quantity': 1}]},
 {'_id': ObjectId('6a3b4c05ac03d6eda1d9ab6e'),
  'customer_id': 3,
  'order_date': '2024-08-15',
  'order_id': 14,
  'products': [{'product_id': 5, 'quantity': 2},
               {'product_id': 8, 'quantity': 1}]}]


## Task 4
Add to your program code to calculate the revenue generated from all orders.

Display the revenue in a tabular format with headers for Order ID and Revenue.

For each order, compute the revenue as the sum of the price of each product multiplied 
by the quantity ordered.

Run your program. <div style="text-align: right">[5]</div>


In [12]:
# Your Answer Here
client = MongoClient('localhost', 27017)
db = client['online_shop']

prod_coll = db['products']
orders_coll = db['orders']

orders = list(orders_coll.find())
print(f'{"Order ID":<10}|{"Revenue":>10}')

for order in orders:
    revenue = 0
    for prod in order['products']:
        price = prod_coll.find_one({'product_id': prod['product_id']})['price']
        revenue += price * prod['quantity']
    print(f'{order["order_id"]:<10}|{revenue:>10}')

Order ID  |   Revenue
1         |       800
2         |      2500
3         |       230
4         |      1280
5         |       350
6         |      2000
7         |      2180
8         |       400
9         |      5650
10        |      1700
11        |      2260
12        |      4400
13        |      2400
14        |       660
15        |      2250



---

# 4. [ASRJC/2024/PRELIM/P2/Q3]
QuickBuy is an online retailer that wants to improve its customer shopping experience. Customers can leave reviews and ratings for products they have purchased.

QuickBuy needs to keep track of its data using both a relational database (SQLite) and a NoSQL database (MongoDB) and object-oriented programming techniques.

The following shows the relational schema for the SQLite database `quickbuy.db` with three tables:
- `customers (customer_id, customer_name, email)`
- `products (product_id, product_name, category)`
- `orders (customer_id#, product_id#, order_date, quantity)`
`#` denotes foreign key


## Task 1
Write program code to create the three tables for the relational database. <div style="text-align: right">[6]</div>


In [1]:
# Your Answer Here
import sqlite3

with sqlite3.connect('quickbuy.db') as conn:
    cursor = conn.cursor()

    cursor.executescript('''
        CREATE TABLE IF NOT EXISTS customers (
                         customer_id TEXT PRIMARY KEY NOT NULL,
                         customer_name TEXT NOT NULL,
                         email TEXT NOT NULL
                         );
        
        CREATE TABLE IF NOT EXISTS products (
                         product_id TEXT PRIMARY KEY NOT NULL,
                         product_name TEXT NOT NULL,
                         category TEXT NOT NULL
                        );
        
        CREATE TABLE IF NOT EXISTS orders (
                         customer_id TEXT NOT NULL,
                         product_id TEXT NOT NULL,
                         order_date TEXT NOT NULL,
                         quantity INT NOT NULL,

                         FOREIGN KEY (customer_id) REFERENCES customers(customer_id),
                         FOREIGN KEY (product_id) REFERENCES products(product_id)
                         );
    ''')

    conn.commit()

## Task 2
Write program code to insert records from `customers.txt`, `products.txt` and `orders.txt` into the three tables. <div style="text-align: right">[5]</div>


In [2]:
# Your Answer Here
import csv
import sqlite3

with sqlite3.connect('quickbuy.db') as conn:
    cursor = conn.cursor()

    with open('./resources/customers.txt') as f:
        data = [i for i in csv.reader(f)]
        for item in data:
            cursor.execute('INSERT INTO customers(customer_id, customer_name, email) VALUES (?, ?, ?)', item)
    
    with open('./resources/products.txt') as f:
        data = [i for i in csv.reader(f)]
        for item in data:
            cursor.execute('INSERT INTO products(product_id, product_name, category) VALUES (?, ?, ?)', item)

    with open('./resources/orders.txt') as f:
        data = [i for i in csv.reader(f)]
        for item in data:
            cursor.execute('INSERT INTO orders(customer_id, product_id, order_date, quantity) VALUES (?, ?, ?, ?)', item)
    
    conn.commit()

## Task 3
Write program code to get and output the total quantity of products ordered by each customer. <div style="text-align: right">[3]</div>


In [8]:
# Your Answer Here
import sqlite3

with sqlite3.connect('quickbuy.db') as conn:
    cursor = conn.cursor()

    cursor.execute('SELECT customer_id FROM customers')
    customer_ids = cursor.fetchall()

    for id in customer_ids:
        cursor.execute('SELECT quantity FROM orders WHERE customer_id = ?', id)
        quantity = sum([i[0] for i in cursor.fetchall()])
        print(f'{id[0]}:{quantity}')

C1:4
C10:3
C2:3
C3:3
C4:3
C5:3
C6:4
C7:4
C8:3
C9:3


## Task 4
Write program code to get and output the number of orders and total quantity for each product category for orders made in the year 2023. <div style="text-align: right">[4]</div>


In [14]:
# Your Answer Here
import sqlite3

with sqlite3.connect('quickbuy.db') as conn:
    cursor = conn.cursor()

    cursor.execute('SELECT o.quantity, p.category FROM orders as o JOIN products as p ON p.product_id = o.product_id WHERE o.order_date LIKE "2023-%"')
    orders = cursor.fetchall()
    print(f'n_orders: {len(orders)}')
    
    cats = {}
    for i in orders:
        if not cats.get(i[1]):
            cats[i[1]] = i[0]
        else:
            cats[i[1]] += i[0]

    print(f'Categories:')
    print(cats)

n_orders: 0
Categories:
{}


## Task 5
Write program code to insert data from reviews.json to a quickbuy MongoDB database reviews collection. <div style="text-align: right">[3]</div>


In [15]:
# Your Answer Here
import json
from pymongo import MongoClient

client = MongoClient('localhost', 27017)
db = client['quickbuy']
coll = db['reviews']

with open('./resources/reviews.json') as f:
    coll.insert_many(json.load(f))

## Task 6
Write program code to find and output all highly rated products (at least rating 4) with qualitative reviews. <div style="text-align: right">[3]</div>


In [ ]:
# Your Answer Here
from pymongo import MongoClient

client = MongoClient('localhost', 27017)
db = client['quickbuy']
coll = db['reviews']

products = coll.find({'rating': {'$gte': 4}})
print([i['product_name'] for i in list(products) if i.get('review_text')])

['Ultra HD Smart TV', 'Ergonomic Office Chair']


## Task 7
Write program code to find and output all the reviews for product P5 and calculate and output its average rating correct to 1 decimal place. <div style="text-align: right">[4]</div>

In [24]:
# Your Answer Here
from pymongo import MongoClient

client = MongoClient('localhost', 27017)
db = client['quickbuy']
coll = db['reviews']

products = list(coll.find({'product_id': 'P5'}))
avg = sum([i['rating'] for i in products]) / len([i['rating'] for i in products])
print(f'{avg}')

4.5



---

# 5. [JPJC/2022/PRELIM/P2/Q3]
Style Theory decides to manage its rental using a NoSQL database instead. 
Information about the items is stored in the JSON file `RENTAL.JSON`. 
The following fields are recorded: 
- catalogue number, 
- brand, 
- category (i.e.: apparel or bag), 
- daily rental fee, 
- size (for apparel only), 
- customer’s email 
- start date, and 
- end date.


## Task 1 
Write program code to import the information from the JSON file into a MongoDB database. Save the information under the Rental collection in the `StyleTheory` database. Ensure that the collection only stores the information from the JSON file. 
<div style="text-align: right">[2]</div>

Save your program code as 
```text
TASK3_<your class>_<your name>.py 
```


In [26]:
# Your Answer Here
from pymongo import MongoClient
import json

client = MongoClient('localhost', 27017)
db = client['StyleTheory']
coll = db['Rental']

with open('./resources/RENTAL.json') as f:
    coll.insert_many(json.load(f))

print(list(coll.find()))

[{'_id': ObjectId('6a3b58fa25cbede4229510dd'), 'catalogueNumber': 1000, 'brand': 'Fendi', 'category': 'apparel', 'rental': 40, 'size': 6, 'email': 'marychia@gmail.com', 'startDate': '2022-01-03', 'endDate': '2022-01-05'}, {'_id': ObjectId('6a3b58fa25cbede4229510de'), 'catalogueNumber': 1371, 'brand': 'Mc Queen', 'category': 'apparel', 'rental': 45, 'size': 4, 'email': 'lee1969rocks@gmail.com', 'startDate': '2022-01-03', 'endDate': '2022-01-09'}, {'_id': ObjectId('6a3b58fa25cbede4229510df'), 'catalogueNumber': 1266, 'brand': 'Chanel', 'category': 'bag', 'rental': 75, 'email': 'lee1969rocks@gmail.com', 'startDate': '2022-01-03', 'endDate': '2022-01-09'}, {'_id': ObjectId('6a3b58fa25cbede4229510e0'), 'catalogueNumber': 1182, 'brand': 'Mc Queen', 'category': 'apparel', 'rental': 80, 'size': 8, 'email': 'feliciaxoxo@yahoo.com', 'startDate': '2022-01-05', 'endDate': '2022-01-08'}, {'_id': ObjectId('6a3b58fa25cbede4229510e1'), 'catalogueNumber': 1375, 'brand': 'Gucci', 'category': 'bag', 'ren

## Task 2 
Write program code for a user to insert information of a new rental or update the end date of an existing rental. Display the following user menu for the user:

```text
Style Theory 
Option 1 – Insert new rental 
Option 2 – Update existing rental 
Option 3 – Quit 
Enter your option (1, 2, or 3): 
```

If Option 1 is entered, get user input for all relevant fields and insert the new document to the collection. Size data input is applicable for apparel only. 

Otherwise, if Option 2 is entered, get user input for the catalogue number, start date and new end date and update the end date of the existing document. 

Run your program to insert the following documents: 

|Field |Document 1| Document 2
|---|---|--- 
|Catalogue Number| 1400 |1375 
|Brand| Trioon |Gucci 
|Category| Apparel| Bag 
|Daily Rental Fee| 20 |50 
|Size |8 |
|Customer Email| liu@gmail.com bliu@gmail.com 
|Start Date| 2022-01-10 b2022-01-10 
|End Date| 2022-01-14 b2022-01-14 

Run your program to update the following document: 

|Field |Document 1| Document 2
|---|---|--- 
|Catalogue Number| 1371| 1266 
|Start Date |2022-01-03| 2022-01-03 
|End Date |2022-01-05| 2022-01-05 

<div style="text-align: right">[10]</div>

Add your program code to 
```text
TASK3_<your class>_<your name>.py 
```


In [29]:
# Your Answer Here
from pymongo import MongoClient

client = MongoClient('localhost', 27017)
db = client['StyleTheory']
coll = db['Rental']

while True:
    print('''
        Style Theory
        Option 1 - Insert new rental
        Option 2 - Update existing rental
        Option 3 - Quit
        
        Enter your option (1, 2, or 3):
    ''')
    ans = input()
    if ans == '3':
        print('Exitted')
        break

    if ans == '1':
        doc = {
            "catalogueNumber": input("Catalogue Number: "),
            "brand": input("Brand: "),
            "category": input("Category: "),
            "rental": input("Daily Rental Fee: "),
            "email": input("Customer Email: "),
            "startDate": input("Start Date: "),
            "endDate": input("End Date: ")
        }

        if doc["category"].lower() == "apparel":
            doc["size"] = input("Size: ")

        coll.insert_one(doc)
    
    elif ans == '2':
        doc = {
            "catalogueNumber": input("Catalogue Number: "),
            "startDate": input("Start Date: ")
        }
        end = input("New End Date: ")

        coll.update_one(doc, {'$set': {'endDate': end}})

    print('Done!')


        Style Theory
        Option 1 - Insert new rental
        Option 2 - Update existing rental
        Option 3 - Quit
        
        Enter your option (1, 2, or 3):
    
Done!

        Style Theory
        Option 1 - Insert new rental
        Option 2 - Update existing rental
        Option 3 - Quit
        
        Enter your option (1, 2, or 3):
    
Done!

        Style Theory
        Option 1 - Insert new rental
        Option 2 - Update existing rental
        Option 3 - Quit
        
        Enter your option (1, 2, or 3):
    
Done!

        Style Theory
        Option 1 - Insert new rental
        Option 2 - Update existing rental
        Option 3 - Quit
        
        Enter your option (1, 2, or 3):
    
Exitted


## Task 3 
Write a procedure `display_all` that will display all the information in the `Rental` collection. Output `“NA”` for size if the document does not contain size field. Include an additional column to display the total amount payable (i.e.: daily rental fee multiplied by the number of days rented). 

You may refer to the following Python code to calculate the difference in the number of days between two dates: 

```text
from datetime import date 
f_date = date(2014, 7, 2) 
l_date = date(2014, 7, 11) 
delta = l_date - f_date 
print(delta.days) 
```
Run the `display_all` procedure. 
<div style="text-align: right">[8]</div>

Add your program code to 
```text
TASK3_<your class>_<your name>.py
```

In [49]:
# Your Answer Here
from pymongo import MongoClient
from datetime import date

def display_all():
    client = MongoClient('localhost', 27017)
    db = client['StyleTheory']
    coll = db['Rental']

    data = coll.find()

    print(f"{'Catalogue Number':<20}|{'Brand':<20}|{'Category':<20}|{'Rental Fee':<20}|{'Size':<20}|{'Email':<30}|{'Start Date':<20}|{'End Date':<20}|{'Amount Payable':<20}")
    for doc in data:
        print(f"{doc.get('catalogueNumber'):<20}|{doc.get('brand'):<20}|{doc.get('category'):<20}|{doc.get('rental'):<20}|{doc.get('size') or 'NA':<20}|{doc.get('email'):<30}|{doc.get('startDate'):<20}|{doc.get('endDate'):<20}|", end='')

        f_date = date(*[int(i) for i in doc['startDate'].split('-')])
        l_date = date(*[int(i) for i in doc['endDate'].split('-')])
        delta = l_date - f_date
        print(delta.days*int(doc.get('rental')))

display_all()

Catalogue Number    |Brand               |Category            |Rental Fee          |Size                |Email                         |Start Date          |End Date            |Amount Payable      
1000                |Fendi               |apparel             |40                  |6                   |marychia@gmail.com            |2022-01-03          |2022-01-05          |80
1371                |Mc Queen            |apparel             |45                  |4                   |lee1969rocks@gmail.com        |2022-01-03          |2022-01-09          |270
1266                |Chanel              |bag                 |75                  |NA                  |lee1969rocks@gmail.com        |2022-01-03          |2022-01-09          |450
1182                |Mc Queen            |apparel             |80                  |8                   |feliciaxoxo@yahoo.com         |2022-01-05          |2022-01-08          |240
1375                |Gucci               |bag                 |50         


---

# 6. [DHS/2022/PRELIM/P2/Q4]
## Task 1 
NDP administrators help in selecting suitable applicants who want to march in NDP contingents.

Write a program to help a NDP administrator insert data into a MongoDB database `applicants` under the collection `contingents`. The data is provided for you in `contingents.json` as well as in the table below where the first row are headers for the fields.

|name| medical_history| citizenship| height| suitability
|---|---|---|---|---
|Shen Rui Lin| heart issues, asthma, eczema| Singaporean| 1.8 |Pending
|Chee Jun Ming| G6PD, heat exhaustion| PR| 1.4 |Pending
|Yeung Jun Feng| heat exhaustion, sweaty palms| Foreigner| 1.7 |Pending
|Deng Kai De| asthma| Singaporean| |Pending

The `medical_history` field is an array of medical conditions and height is in metres.

Insert the above data and print out all documents in the collection contingents.
<div style="text-align: right">[6]</div>


In [ ]:
# Your Answer Here
from pymongo import MongoClient
import json

client = MongoClient('localhost', 27017)
db = client['applicants']
coll = db['contingents']

with open('./resources/contingents.json') as f:
    coll.insert_many(json.load(f))

print(list(coll.find()))

[{'_id': ObjectId('6a3b676625cbede42295110b'), 'citizenship': 'Singaporean', 'height': 1.8, 'medical_history': ['heart issues', 'asthma', 'eczema'], 'name': 'Shen Rui Lin', 'suitability': 'Pending'}, {'_id': ObjectId('6a3b676625cbede42295110c'), 'citizenship': 'PR', 'height': 1.4, 'medical_history': ['G6PD', 'heat exhaustion'], 'name': 'Chee Jun Ming', 'suitability': 'Pending'}, {'_id': ObjectId('6a3b676625cbede42295110d'), 'citizenship': 'Foreigner', 'height': 1.7, 'medical_history': ['heat exhaustion', 'sweaty palms'], 'name': 'Yeung Jun Feng', 'suitability': 'Pending'}, {'_id': ObjectId('6a3b676625cbede42295110e'), 'citizenship': 'Singaporean', 'medical_history': ['asthma'], 'name': 'Deng Kai De', 'suitability': 'Pending'}]


## Task 2
Write code to query the database for all documents of Singaporean applicants who do not have `“heart issues”` in their medical history and who are taller than 1.5m (if they submitted a height value).

Update the `suitability` of these document(s) to `“High”`.

Finally, print out **all** documents in `contingents`.
<div style="text-align: right">[5]</div>

In [57]:
# Your Answer Here
from pymongo import MongoClient

client = MongoClient('localhost', 27017)
db = client['applicants']
coll = db['contingents']

coll.update_many({'medical_history': {'$ne': 'heart issues'},
                  'height': {'$gt': 1.5}
                  },
                  
                  {
                      '$set': {'suitability': 'High'}
                  })

list(coll.find())

[{'_id': ObjectId('6a3b676625cbede42295110b'),
  'citizenship': 'Singaporean',
  'height': 1.8,
  'medical_history': ['heart issues', 'asthma', 'eczema'],
  'name': 'Shen Rui Lin',
  'suitability': 'Pending'},
 {'_id': ObjectId('6a3b676625cbede42295110c'),
  'citizenship': 'PR',
  'height': 1.4,
  'medical_history': ['G6PD', 'heat exhaustion'],
  'name': 'Chee Jun Ming',
  'suitability': 'Pending'},
 {'_id': ObjectId('6a3b676625cbede42295110d'),
  'citizenship': 'Foreigner',
  'height': 1.7,
  'medical_history': ['heat exhaustion', 'sweaty palms'],
  'name': 'Yeung Jun Feng',
  'suitability': 'High'},
 {'_id': ObjectId('6a3b676625cbede42295110e'),
  'citizenship': 'Singaporean',
  'medical_history': ['asthma'],
  'name': 'Deng Kai De',
  'suitability': 'Pending'}]


---

# 7. [RVHS/2021/PRELIM/P2/Q3]


## Task 1
Write program code to read the csv file `"health_facilities.csv"` and insert all information in the file as documents into a NoSQL MongoDB database called `"Health"` with one collection called `"facilities"`. The `"_id"` of the documents in the database should start from `1`, `2`, `3` and `4` etc. The correct data type of each field is expected to be inserted into the database. 

<div style="text-align: right">[10]</div>


In [64]:
# Your Answer Here
import csv
from pymongo import MongoClient

client = MongoClient('localhost', 27017)
db = client['Health']
coll = db['facilities']

with open('./resources/health_facilities.csv') as f:
    data = [i for i in csv.reader(f)]

headers = data[0]
query = []
for i, entry in enumerate(data[1:]):
    doc = {'_id': i+1}
    for j in range(len(entry)):
        doc[headers[j]] = int(entry[j]) if entry[j].isdigit() else entry[j]

    query.append(doc)

coll.insert_many(query)

InsertManyResult([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180], acknowledged=True)

## Task 2
- a) Write a MongoDB Pymongo query to retrieve all public acute hospital documents with their corresponding number of beds more than 7200. 
<div style="text-align: right">[4]</div>

- b) Write program code to bubble sort the results retrieved in **Task 3.2** a) according to the average number of beds per facility. Then, display the top 3 years which has the highest average number of beds per facility using the format below.
 ```text
 The three years that have the highest average number of beds per facility are: _____, _____ and _____.
 ```
<div style="text-align: right">[7]</div>


In [74]:
# Your Answer Here
from pymongo import MongoClient

client = MongoClient('localhost', 27017)
db = client['Health']
coll = db['facilities']

# a)
docs = list(coll.find({
    'no_beds': {'$gt': 7200},
    'public_private': 'Public'
}))

#b)
for i in range(len(docs)-1):
    for j in range(len(docs)-1-i):
        avg_j_1 = docs[j+1]['no_beds'] / docs[j+1]['no_of_facilities']
        avg_j = docs[j]['no_beds'] / docs[j]['no_of_facilities']

        if avg_j > avg_j_1:
            docs[j], docs[j+1] = docs[j+1], docs[j]

print(f'The three years that have the highest average number of beds per facility are: {docs[-1]["year"]}, {docs[-2]["year"]} and {docs[-3]["year"]}')

The three years that have the highest average number of beds per facility are: 2020, 2017 and 2014


## Task 3
- a) Write a MongoDB Pymongo query and program code to display all `"_id"`s of Not-for-Profit health facilities documents that have no facility. 
<div style="text-align: right">[3]</div>

- b) Write MongoDB Pymongo code to update the fields "`no_of_facilities`" 
and "`no_beds`" of only 3 documents retrieved in Task 3.3 a) to `1` and a 
random number from `10` to `20` respectively. 
<div style="text-align: right">[6]</div>

In [80]:
# Your Answer Here
from pymongo import MongoClient
import random

client = MongoClient('localhost', 27017)
db = client['Health']
coll = db['facilities']

# a)
docs = list(coll.find({
    'no_of_facilities': 0,
    'public_private': 'Not-for-Profit'
}))
ids = [i['_id'] for i in docs]
print(ids[:3])

# b)
coll.update_many({'_id': {'$in': ids[:3]}}, 
                 {
                     '$set': {'no_of_facilities': 1, 'no_beds': random.randint(10, 20)}
                 }
                 )

[5, 20, 35]


UpdateResult({'n': 3, 'nModified': 3, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)


---

# 8. [JPJC/2021/PRELIM/P2/Q4]
JP Mobile sells mobile phones and manages its inventory using a NoSQL database. Information about the mobile phones is stored in the JSON file `items.JSON`.
The following fields are recorded:
- brand of mobile phone,
- model,
- colour(s) available,
- price in dollar,
- quantity in stock.


## Task 1
Write program code to import the information from the JSON file into a MongoDB database. Save the information under the `phone` collection in the `jp_mobile` database. Ensure that the collection only stores the information from the JSON file.

Save your program code as
```text
TASK4_<your name>_<class>_<index number>.py 
```
<div style="text-align: right">[4]</div>


In [ ]:
# Your Answer Here
import json
from pymongo import MongoClient

client = MongoClient('localhost', 27017)
db = client['jp_mobile']

## Task 2
The shop decides to include one or more free gifts for new batches of mobile phones it sells. 

Write program code for a user to insert information of a mobile phone by getting user input of the following: brand, model, colour, price, quantity, free gift(s). 

Your code should allow user to input one or more free gifts.

If a phone’s `brand`, `model` and `colour` already exists, add the new quantity to the existing quantity in the database, and replace the existing price with the new price.

Run your program and insert the following 2 documents:

No.| Brand |Model| Colour| Price| Quantity| Free gift(s)
---|---|---|---|---|---|---|
1 |orange |22 |black |900 |11 |power bank
2 |solo |A33 |red |1300 |7 |- power bank -earbuds

Add your program code to
```text
TASK4_<your name>_<class>_<index number>.py 
```
<div style="text-align: right">[7]</div>


In [ ]:
# Your Answer Here

## Task 3
Write a function `display_all` that will display all the information in the `phone` collection under these fields: `brand`, `model`, `colour`, `price`, `quantity`, `free gift(s)`.

If no free gift comes with the phone, print a `None` statement. 

Include a final statement that shows the total number of documents in the collection.

Run the `display_all` function to show your output.

Add your program code to
```text
TASK4_<your name>_<class>_<index number>.py
```
<div style="text-align: right">[6]</div>


In [ ]:
# Your Answer Here

## Task 4
The shop uses a web browser to display the database content. The manager wants to filter the mobile phones by `brand` and display the results in a web browser.

Write additional Python code and the necessary files to create a web application that
- receives a `brand` string from a HTML form, then
- creates and returns a HTML document that enables the web browser to display an ordered list of mobile phones sorted by `price`.

For each document, the web page should include the:
- brand as the heading
- model, 
- colour, 
- price

Save your program as
```text
TASK4_4_<your name>_<class>_<index number>.py
```

with any additional files / sub-folders as needed in a folder named
```text
TASK4_4_<your name>_<class>_<index number>
```

Run the web application and input brand as `‘solo’` in the webpage.

Save the output of the program as
```text
TASK4_4_<your name>_<class>_<index number>.html 
```
<div style="text-align: right">[12]</div>


In [ ]:
# Your Answer Here

---